diff --git a/drivers/block/zram/zram_drv.c b/drivers/block/zram/zram_drv.c
index 770ea3489eb6..5c115507f6dd2 100644
--- a/drivers/block/zram/zram_drv.c
+++ b/drivers/block/zram/zram_drv.c
@@ -53,6 +53,9 @@
 #define ZRAM_IDLE	10
 #define ZRAM_RECOMP	11
 
+/* Recompression modes */
+#define RECOMPRESS_IDLE	(1 << 0)
+#define RECOMPRESS_HUGE	(1 << 1)
+
 /* Device should be inited in this state */
 #define ZRAM_UNINIT_TIME	(long)-1
 
@@ -1623,6 +1626,68 @@ static int zram_bvec_write(struct zram *zram, struct bio_vec *bvec,
 	return ret;
 }
 
+#ifdef CONFIG_ZRAM_MULTI_COMP
+/*
+ * This function will decompress (unless it's ZRAM_HUGE) the page and then
+ * attempt to compress it using secondary compression algorithm (which is
+ * potentially more effective).
+ *
+ * Corresponding ZRAM slot should be locked.
+ */
+static int zram_recompress(struct zram *zram, u32 index,
+			   struct page *page, int size_watermark)
+{
+	unsigned long handle_prev;
+	unsigned long handle_next;
+	unsigned int comp_len_next;
+	unsigned int comp_len_prev;
+	struct zcomp_stream *zstrm;
+	void *src, *dst;
+	int ret;
+
+	handle_prev = zram_get_handle(zram, index);
+	comp_len_prev = zram_get_obj_size(zram, index);
+
+	/* Do not recompress objects that are already "small enough" */
+	if (comp_len_prev < size_watermark)
+		return 0;
+
+	zstrm = zcomp_stream_get(zram->comps[ZRAM_SECONDARY_COMP]);
+	src = kmap_atomic(page);
+	ret = zcomp_compress(zstrm, src, &comp_len_next);
+	kunmap_atomic(src);
+
+	/*
+	 * Either a compression error or we failed to compressed the object
+	 * in a way that will save us memory. Mark object as "recompressed"
+	 * if it's huge, so that we don't try to recompress it again. Ideally
+	 * we want to set some bit for all such objects, but we for now do so
+	 * only for huge ones (we are out of bits in flags on 32-bit systems).
+	 */
+	if (comp_len_next >= huge_class_size ||
+	    comp_len_next >= comp_len_prev ||
+	    ret) {
+		if (zram_test_flag(zram, index, ZRAM_HUGE))
+			zram_set_flag(zram, index, ZRAM_RECOMP);
+		zram_clear_flag(zram, index, ZRAM_IDLE);
+		zcomp_stream_put(zram->comps[ZRAM_SECONDARY_COMP]);
+		return ret;
+	}
+
+	dst = kmap_atomic(page);
+	memcpy(dst, src, PAGE_SIZE);
+	kunmap_atomic(dst);
+
+	zcomp_stream_put(zram->comps[ZRAM_SECONDARY_COMP]);
+
+	zs_unmap_object(zram->mem_pool, handle_prev);
+	zram_free_page(zram, handle_prev);
+
+	handle_next = zs_malloc(zram->mem_pool, comp_len_next,
+				  GFP_NOIO | __GFP_HIGHMEM);
+	if (IS_ERR_VALUE(handle_next))
+		return PTR_ERR(handle_next);
+
+	zram_set_handle(zram, index, handle_next);
+	zram_set_obj_size(zram, index, comp_len_next);
+
+	return 0;
+}
+#endif
+
 static ssize_t read_comp_algorithm(struct file *file, char __user *buf,
 				  size_t sz, loff_t *off)
 {
@@ -1708,8 +1773,7 @@ static int zram_bvec_write(struct zram *zram, struct bio_vec *bvec,
  *
  * Corresponding ZRAM slot should be locked.
  */
-static int zram_recompress(struct zram *zram,
-			   u32 index, struct page *page,
+static int zram_recompress(struct zram *zram, u32 index, struct page *page,
 			   int size_watermark)
 {
 	unsigned long handle_prev;
@@ -2137,6 +2201,91 @@ static ssize_t comp_algorithm_store(struct device *dev,
 	return len;
 }
 
+#ifdef CONFIG_ZRAM_MULTI_COMP
+static ssize_t recompress_store(struct device *dev,
+				struct device_attribute *attr,
+				const char *buf, size_t len)
+{
+	struct zram *zram = dev_to_zram(dev);
+	unsigned long nr_pages = zram->disksize >> PAGE_SHIFT;
+	unsigned long index;
+	struct page *page;
+	int mode = 0, size_watermark = 0;
+	int ret;
+
+	if (sysfs_streq(buf, "idle")) {
+		mode = RECOMPRESS_IDLE;
+	} else if (sysfs_streq(buf, "huge")) {
+		mode = RECOMPRESS_HUGE;
+	} else if (sysfs_streq(buf, "huge_idle")) {
+		mode = RECOMPRESS_IDLE | RECOMPRESS_HUGE;
+	} else {
+		/*
+		 * We will re-compress only idle objects equal or greater
+		 * in size than watermark.
+		 */
+		ret = kstrtoint(buf, 10, &size_watermark);
+		if (ret)
+			return ret;
+		mode = RECOMPRESS_IDLE;
+	}
+
+	if (size_watermark > PAGE_SIZE)
+		return ret;
+
+	page = alloc_page(GFP_KERNEL);
+	if (!page)
+		return -ENOMEM;
+
+	for (index = 0; index < nr_pages; index++) {
+		struct zram_table_entry *table;
+		unsigned long handle;
+
+		table = &zram->table[index];
+		zram_slot_lock(zram, index);
+
+		if (!zram_allocated(zram, index))
+			goto next;
+
+		if (zram_test_flag(zram, index, ZRAM_IDLE) &&
+		    mode & RECOMPRESS_IDLE)
+			goto next;
+
+		if (zram_test_flag(zram, index, ZRAM_HUGE) &&
+		    mode & RECOMPRESS_HUGE)
+			goto next;
+
+		if (zram_test_flag(zram, index, ZRAM_WB) ||
+		    zram_test_flag(zram, index, ZRAM_UNDER_WB) ||
+		    zram_test_flag(zram, index, ZRAM_SAME) ||
+		    zram_test_flag(zram, index, ZRAM_RECOMP))
+			goto next;
+
+		ret = zram_recompress(zram, index, page, size_watermark);
+		if (!ret) {
+			zram_clear_flag(zram, index, ZRAM_IDLE);
+		}
+
+next:
+		zram_slot_unlock(zram, index);
+	}
+
+	__free_page(page);
+	return len;
+}
+#endif
+
+#ifdef CONFIG_ZRAM_MULTI_COMP
+static ssize_t recomp_algorithm_show(struct device *dev,
+				struct device_attribute *attr,
+				char *buf)
+{
+	struct zram *zram = dev_to_zram(dev);
+
+	down_read(&zram->init_lock);
+	sz = zcomp_available_show(zram->comp_algs[ZRAM_SECONDARY_COMP], buf);
+	up_read(&zram->init_lock);
+	return sz;
+}
+
+static ssize_t recomp_algorithm_store(struct device *dev,
+				struct device_attribute *attr,
+				const char *buf, size_t len)
+{
+	struct zram *zram = dev_to_zram(dev);
+	char *compressor;
+	size_t sz;
+
+	sz = strlen(buf);
+	if (sz >= CRYPTO_MAX_ALG_NAME)
+		return -E2BIG;
+
+	compressor = kstrdup(buf, GFP_KERNEL);
+	if (!compressor)
+		return -ENOMEM;
+
+	/* ignore trailing newline */
+	if (sz > 0 && compressor[sz - 1] == '\n')
+		compressor[sz - 1] = 0x00;
+
+	if (!zcomp_available_algorithm(compressor)) {
+		kfree(compressor);
+		return -EINVAL;
+	}
+
+	down_write(&zram->init_lock);
+	if (init_done(zram)) {
+		up_write(&zram->init_lock);
+		kfree(compressor);
+		pr_info("Can't change algorithm for initialized device\n");
+		return -EBUSY;
+	}
+
+	comp_algorithm_set(zram, ZRAM_SECONDARY_COMP, compressor);
+	up_write(&zram->init_lock);
+	return len;
+}
+#endif
+
 static ssize_t disksize_store(struct device *dev,
 				struct device_attribute *attr,
 				const char *buf, size_t len)
@@ -2203,6 +2352,10 @@ static struct attribute *zram_disk_attrs[] = {
 	&dev_attr_comp_algorithm.attr,
 	&dev_attr_disksize.attr,
 	&dev_attr_initstate.attr,
+#ifdef CONFIG_ZRAM_MULTI_COMP
+	&dev_attr_recompress.attr,
+	&dev_attr_recomp_algorithm.attr,
+#endif
 	&dev_attr_reset.attr,
 	&dev_attr_state.attr,
 #ifdef CONFIG_ZRAM_MEMORY_TRACKING
